mlflow/pyfunc/__init__.py

```diff
@@ -266,7 +270,117 @@ def _load_model_env(path):
     return _get_flavor_configuration(model_path=path, flavor_name=FLAVOR_NAME).get(ENV, None)
 
 
-def load_model(model_uri, suppress_warnings=True):
+PyFuncOutput = Union[pandas.DataFrame, pandas.Series, np.ndarray, list]
+
+
+class PyFuncModel(object):
+    """
+    MLflow 'python function' model.
+
+    Wrapper around model implementation and metadata. This class is not meant to be constructed
+    directly. Instead, instances of this class are constructed and returned from
+    py:func:`mlflow.pyfunc.load_model`.
+
+    ``model_impl`` can be any Python object that implements the `Pyfunc interface
+    <https://mlflow.org/docs/latest/python_api/mlflow.pyfunc.html#pyfunc-inference-api>`_, and is
+    by invoking the model's ``loader_module``.
+
+    ``model_meta`` contains model metadata loaded from the MLmodel file.
+    """
+
+    def __init__(self, model_meta: Model, model_impl: Any):
+        if not hasattr(model_impl, "predict"):
+            raise MlflowException("Model implementation is missing required predict method.")
+        self._model_meta = model_meta
+        self._model_impl = model_impl
+
+    def predict(self, data: pandas.DataFrame) -> PyFuncOutput:
+        """
+        Generate model predictions.
+        :param data: Model input as pandas.DataFrame.
+        :return: Model predictions as one of pandas.DataFrame, pandas.Series, numpy.ndarray or list.
+        """
+        input_schema = self._model_meta.get_input_schema()
+
+        if input_schema is not None:
+            def convert_type(name, values: pandas.Series, t: DataType):
+                if values.dtype == np.object and t not in (DataType.binary, DataType.string):
+                    values = values.infer_objects()
+
+                if values.dtype in (t.to_pandas(), t.to_numpy()):
+                    return values
+
+                if t == DataType.binary and values.dtype.kind == t.binary.to_numpy().kind:
+                    #  NB: bytes in numpy have variable itemsize and would fail the upcast check.
```

Comment:
This seems to be a forward reference to some logic. Maybe we could say earlier what we plan to do given an input schema?

Reply:
I am not sure what do you mean? The reason this is necessary here is that (and tbh I don't think this happens very often) if you add a column encoded as numpy bytes, the data type will be S<itemsize> where itemsize is the length of the longest item. The itemsize of our numpy type for binary is 0, so the upcast logic would not work here.

Reply:
Right, sorry, I just meant that it wasn't clear at this point in the code that we were going to do an upcast later. You never mention what the code looks like, so until you read 30 lines later, you don't know why this "would fail the upcast check".